File: sdk/ruby/test/test_interchange.rb
# frozen_string_literal: true

require_relative "test_helper"
require "json"

class TestInterchange < Minitest::Test
  def test_export_string_schema
    s = AnyVali.string.min_length(1).max_length(100)
    doc = AnyVali.export(s)
    assert_equal "1.0", doc["anyvaliVersion"]
    assert_equal "1", doc["schemaVersion"]
    assert_equal "string", doc["root"]["kind"]
    assert_equal 1, doc["root"]["minLength"]
    assert_equal 100, doc["root"]["maxLength"]
  end

  def test_export_number_schema
    s = AnyVali.number.min(0).max(100)
    doc = AnyVali.export(s)
    assert_equal "number", doc["root"]["kind"]
    assert_equal 0, doc["root"]["min"]
    assert_equal 100, doc["root"]["max"]
  end

  def test_export_object_schema
    s = AnyVali.object(
      properties: { "name" => AnyVali.string, "age" => AnyVali.int_ },
      required: ["name", "age"]
    )
    doc = AnyVali.export(s)
    assert_equal "object", doc["root"]["kind"]
    assert_equal "string", doc["root"]["properties"]["name"]["kind"]
    assert_equal "int", doc["root"]["properties"]["age"]["kind"]
    assert_equal ["name", "age"], doc["root"]["required"]
  end

  def test_export_portable_fails_with_custom
    s = AnyVali.string.refine { |v, p| nil }
    assert_raises(AnyVali::ValidationError) { AnyVali.export(s) }
  end

  def test_export_extended_with_custom
    s = AnyVali.string.refine { |v, p| nil }
    doc = AnyVali.export(s, mode: :extended)
    assert_equal "string", doc["root"]["kind"]
  end

  def test_import_string_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "string", "minLength" => 3 },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal "hello", schema.parse("hello")
    result = schema.safe_parse("ab")
    assert result.failure?
    assert_equal "too_small", result.issues.first.code
  end

  def test_import_number_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "number", "min" => 10 },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal 15, schema.parse(15)
    result = schema.safe_parse(5)
    assert result.failure?
  end

  def test_import_int_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "int", "min" => 1, "max" => 10 },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal 5, schema.parse(5)
  end

  def test_import_bool_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "bool" },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal true, schema.parse(true)
  end

  def test_import_null_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "null" },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_nil schema.parse(nil)
  end

  def test_import_object_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => {
        "kind" => "object",
        "properties" => {
          "name" => { "kind" => "string" },
          "age" => { "kind" => "int" }
        },
        "required" => ["name", "age"],
        "unknownKeys" => "reject"
      },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    result = schema.parse({ "name" => "Alice", "age" => 30 })
    assert_equal({ "name" => "Alice", "age" => 30 }, result)
  end

  def test_import_array_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "array", "items" => { "kind" => "string" } },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal ["a", "b"], schema.parse(["a", "b"])
  end

  def test_import_tuple_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => {
        "kind" => "tuple",
        "elements" => [{ "kind" => "string" }, { "kind" => "int" }]
      },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal ["hello", 42], schema.parse(["hello", 42])
  end

  def test_import_union_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => {
        "kind" => "union",
        "variants" => [{ "kind" => "string" }, { "kind" => "int" }]
      },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal "hello", schema.parse("hello")
    assert_equal 42, schema.parse(42)
  end

  def test_import_record_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "record", "values" => { "kind" => "int" } },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal({ "a" => 1, "b" => 2 }, schema.parse({ "a" => 1, "b" => 2 }))
  end

  def test_import_nullable_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "nullable", "schema" => { "kind" => "string" } },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_nil schema.parse(nil)
    assert_equal "hello", schema.parse("hello")
  end

  def test_import_optional_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => {
        "kind" => "object",
        "properties" => {
          "name" => { "kind" => "optional", "schema" => { "kind" => "string" } }
        },
        "required" => [],
        "unknownKeys" => "reject"
      },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal({}, schema.parse({}))
    assert_equal({ "name" => "Alice" }, schema.parse({ "name" => "Alice" }))
  end

  def test_import_literal_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "literal", "value" => "hello" },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal "hello", schema.parse("hello")
  end

  def test_import_enum_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "enum", "values" => ["red", "green", "blue"] },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal "red", schema.parse("red")
  end

  def test_import_ref_schema
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => {
        "kind" => "object",
        "properties" => {
          "user" => { "kind" => "ref", "ref" => "#/definitions/User" }
        },
        "required" => ["user"],
        "unknownKeys" => "reject"
      },
      "definitions" => {
        "User" => {
          "kind" => "object",
          "properties" => {
            "name" => { "kind" => "string" },
            "age" => { "kind" => "int" }
          },
          "required" => ["name", "age"],
          "unknownKeys" => "reject"
        }
      },
      "extensions" => {}
    }
    schema, context, _defs = AnyVali.import(doc)
    result = schema.safe_parse({ "user" => { "name" => "Alice", "age" => 30 } }, context: context)
    assert result.success?
    assert_equal({ "user" => { "name" => "Alice", "age" => 30 } }, result.value)
  end

  def test_import_from_json_string
    json = '{"anyvaliVersion":"1.0","schemaVersion":"1","root":{"kind":"string"},"definitions":{},"extensions":{}}'
    schema = AnyVali.import_schema(json)
    assert_equal "hello", schema.parse("hello")
  end

  def test_roundtrip_export_import
    original = AnyVali.object(
      properties: {
        "name" => AnyVali.string.min_length(1),
        "age" => AnyVali.int_.min(0)
      },
      required: ["name", "age"]
    )
    doc = AnyVali.export(original)
    reimported = AnyVali.import_schema(doc)
    assert_equal({ "name" => "Alice", "age" => 30 }, reimported.parse({ "name" => "Alice", "age" => 30 }))
  end

  def test_import_coercion
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "int", "coerce" => "string->int" },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal 42, schema.parse("42")
  end

  def test_import_chained_coercion
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "string", "coerce" => ["trim", "lower"] },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal "hello", schema.parse("  HELLO  ")
  end

  def test_import_intersection
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => {
        "kind" => "intersection",
        "allOf" => [
          { "kind" => "number", "min" => 0 },
          { "kind" => "number", "max" => 100 }
        ]
      },
      "definitions" => {},
      "extensions" => {}
    }
    schema = AnyVali.import_schema(doc)
    assert_equal 50, schema.parse(50)
  end

  def test_import_unsupported_kind
    doc = {
      "anyvaliVersion" => "1.0",
      "schemaVersion" => "1",
      "root" => { "kind" => "custom_thing" },
      "definitions" => {},
      "extensions" => {}
    }
    assert_raises(AnyVali::ValidationError) { AnyVali.import_schema(doc) }
  end

  def test_import_width_types
    %w[int8 int16 int32 int64 uint8 uint16 uint32 uint64 float32 float64].each do |kind|
      doc = {
        "anyvaliVersion" => "1.0",
        "schemaVersion" => "1",
        "root" => { "kind" => kind },
        "definitions" => {},
        "extensions" => {}
      }
      schema = AnyVali.import_schema(doc)
      assert_equal kind, schema.kind
    end
  end

  def test_import_any_unknown_never
    %w[any unknown never].each do |kind|
      doc = {
        "anyvaliVersion" => "1.0",
        "schemaVersion" => "1",
        "root" => { "kind" => kind },
        "definitions" => {},
        "extensions" => {}
      }
      schema = AnyVali.import_schema(doc)
      assert_equal kind, schema.kind
    end
  end
end

File: sdk/cpp/tests/test_interchange.cpp
#include "test_framework.hpp"
#include <anyvali/anyvali.hpp>

using namespace anyvali;
using json = nlohmann::json;

TEST("import: basic string schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "string"}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(imported.root->kind() == SchemaKind::String);
    auto result = parse_document(imported, json("hello"));
    ASSERT(result.success);
}

TEST("import: number with constraints") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "number"}, {"min", 0}, {"max", 100}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json(50)).success);
    ASSERT(!parse_document(imported, json(-1)).success);
}

TEST("import: object schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "object"},
            {"properties", {
                {"name", {{"kind", "string"}}},
                {"age", {{"kind", "int"}}}
            }},
            {"required", {"name", "age"}},
            {"unknownKeys", "reject"}
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported, json::object({{"name", "Alice"}, {"age", 30}}));
    ASSERT(result.success);
}

TEST("import: union schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "union"},
            {"variants", {{{"kind", "string"}}, {{"kind", "int"}}}}
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json("hello")).success);
    ASSERT(parse_document(imported, json(42)).success);
    ASSERT(!parse_document(imported, json(true)).success);
}

TEST("import: intersection schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "intersection"},
            {"allOf", {
                {{"kind", "number"}, {"min", 0}},
                {{"kind", "number"}, {"max", 100}}
            }}
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json(50)).success);
}

TEST("import: nullable schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "nullable"}, {"schema", {{"kind", "string"}}}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json(nullptr)).success);
    ASSERT(parse_document(imported, json("hello")).success);
}

TEST("import: array schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "array"}, {"items", {{"kind", "int"}}}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json::array({1, 2, 3})).success);
}

TEST("import: tuple schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "tuple"}, {"elements", {{{"kind", "string"}}, {{"kind", "int"}}}}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json::array({"hello", 42})).success);
}

TEST("import: record schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "record"}, {"values", {{"kind", "int"}}}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json::object({{"a", 1}, {"b", 2}})).success);
}

TEST("import: literal schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "literal"}, {"value", "hello"}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json("hello")).success);
    ASSERT(!parse_document(imported, json("world")).success);
}

TEST("import: enum schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "enum"}, {"values", {"red", "green", "blue"}}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json("red")).success);
    ASSERT(!parse_document(imported, json("yellow")).success);
}

TEST("import: ref schema with definitions") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "object"},
            {"properties", {
                {"user", {{"kind", "ref"}, {"ref", "#/definitions/User"}}}
            }},
            {"required", {"user"}},
            {"unknownKeys", "reject"}
        }},
        {"definitions", {
            {"User", {
                {"kind", "object"},
                {"properties", {
                    {"name", {{"kind", "string"}}},
                    {"age", {{"kind", "int"}}}
                }},
                {"required", {"name", "age"}},
                {"unknownKeys", "reject"}
            }}
        }},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported,
        json::object({{"user", json::object({{"name", "Alice"}, {"age", 30}})}}));
    ASSERT(result.success);
}

TEST("import: recursive ref") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "ref"}, {"ref", "#/definitions/TreeNode"}}},
        {"definitions", {
            {"TreeNode", {
                {"kind", "object"},
                {"properties", {
                    {"value", {{"kind", "int"}}},
                    {"children", {
                        {"kind", "array"},
                        {"items", {{"kind", "ref"}, {"ref", "#/definitions/TreeNode"}}}
                    }}
                }},
                {"required", {"value", "children"}},
                {"unknownKeys", "reject"}
            }}
        }},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    json tree = {
        {"value", 1},
        {"children", {
            {{"value", 2}, {"children", json::array()}},
            {{"value", 3}, {"children", {
                {{"value", 4}, {"children", json::array()}}
            }}}
        }}
    };
    ASSERT(parse_document(imported, tree).success);
}

TEST("import: coercion config") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "int"}, {"coerce", "string->int"}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported, json("42"));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json(42));
}

TEST("import: chained coercion config") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "string"}, {"coerce", {"trim", "lower"}}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported, json("  HELLO  "));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json("hello"));
}

TEST("import: default value") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "object"},
            {"properties", {
                {"name", {{"kind", "string"}}},
                {"role", {{"kind", "string"}, {"default", "user"}}}
            }},
            {"required", {"name"}},
            {"unknownKeys", "reject"}
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported, json::object({{"name", "Alice"}}));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value["role"], json("user"));
}

TEST("export: round-trip simple schema") {
    AnyValiDocument doc;
    doc.root = string_();
    auto exported = export_schema(doc);
    ASSERT(exported["root"]["kind"] == "string");
    ASSERT(exported["anyvaliVersion"] == "1.0");
}

TEST("export: round-trip object schema") {
    AnyValiDocument doc;
    auto obj = object();
    obj->prop("name", string_());
    obj->prop("age", int_());
    obj->required({"name", "age"});
    doc.root = obj;
    auto exported = export_schema(doc);
    ASSERT(exported["root"]["kind"] == "object");
    ASSERT(exported["root"]["properties"]["name"]["kind"] == "string");
}

TEST("export: round-trip with definitions") {
    AnyValiDocument doc;
    auto user_schema = object();
    user_schema->prop("name", string_());
    user_schema->required({"name"});
    doc.definitions["User"] = user_schema;
    doc.root = ref("#/definitions/User");
    auto exported = export_schema(doc);
    ASSERT(exported["definitions"]["User"]["kind"] == "object");
    ASSERT(exported["root"]["kind"] == "ref");
}

TEST("export: portable mode") {
    AnyValiDocument doc;
    doc.root = string_();
    auto exported = export_schema(doc, ExportMode::Portable);
    ASSERT(exported["extensions"].empty());
}

TEST("export: extended mode with extensions") {
    AnyValiDocument doc;
    doc.root = string_();
    doc.extensions = {{"cpp", {{"custom", true}}}};
    auto exported = export_schema(doc, ExportMode::Extended);
    ASSERT(exported["extensions"]["cpp"]["custom"] == true);
}

TEST("import: all primitive kinds") {
    std::vector<std::string> kinds = {
        "any", "unknown", "never", "null", "bool", "string",
        "number", "int", "float32", "float64",
        "int8", "int16", "int32", "int64",
        "uint8", "uint16", "uint32", "uint64"
    };
    for (const auto& k : kinds) {
        json doc = {
            {"anyvaliVersion", "1.0"},
            {"schemaVersion", "1"},
            {"root", {{"kind", k}}},
            {"definitions", json::object()},
            {"extensions", json::object()}
        };
        auto imported = import_schema(doc);
        ASSERT(kind_to_string(imported.root->kind()) == k);
    }
}

TEST("import: optional schema") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "object"},
            {"properties", {
                {"name", {{"kind", "optional"}, {"schema", {{"kind", "string"}}}}}
            }},
            {"required", json::array()},
            {"unknownKeys", "reject"}
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    ASSERT(parse_document(imported, json::object({{"name", "Alice"}})).success);
    ASSERT(parse_document(imported, json::object()).success);
}

TEST("import: strip unknown keys") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "object"},
            {"properties", {{"name", {{"kind", "string"}}}}},
            {"required", {"name"}},
            {"unknownKeys", "strip"}
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported,
        json::object({{"name", "Alice"}, {"extra", "value"}, {"another", 42}}));
    ASSERT(result.success);
    ASSERT_JSON_EQ(result.value, json::object({{"name", "Alice"}}));
}

TEST("import: unsupported schema kind") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {{"kind", "bogus_xyz"}}},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    bool threw = false;
    try {
        import_schema(doc);
    } catch (...) {
        threw = true;
    }
    ASSERT(threw);
}

TEST("import: missing kind field") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", json::object()},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    bool threw = false;
    try {
        import_schema(doc);
    } catch (...) {
        threw = true;
    }
    ASSERT(threw);
}

TEST("import: null empty root") {
    json doc_empty = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    bool threw1 = false;
    try {
        import_schema(doc_empty);
    } catch (...) {
        threw1 = true;
    }
    ASSERT(threw1);

    json doc_null = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", nullptr},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    bool threw2 = false;
    try {
        import_schema(doc_null);
    } catch (...) {
        threw2 = true;
    }
    ASSERT(threw2);
}

TEST("import: default unknown keys is reject") {
    json doc = {
        {"anyvaliVersion", "1.0"},
        {"schemaVersion", "1"},
        {"root", {
            {"kind", "object"},
            {"properties", {{"name", {{"kind", "string"}}}}},
            {"required", {"name"}}
            // no unknownKeys field -> defaults to reject
        }},
        {"definitions", json::object()},
        {"extensions", json::object()}
    };
    auto imported = import_schema(doc);
    auto result = parse_document(imported,
        json::object({{"name", "Alice"}, {"extra", "value"}}));
    ASSERT(!result.success);
    ASSERT(result.issues[0].code == "unknown_key");
}
